### PR TITLE
fix: msg forwarding when flood publish is on

### PR DESCRIPTION
**Pull request by bochaco**
## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/gossipsub/src/behaviour.rs [hunk 1]
```diff
@@ -661,7 +661,8 @@ where
             && self.forward_msg(&msg_id, raw_message.clone(), None, HashSet::new())?;
 
         let mut recipient_peers = HashSet::new();
-        if let Some(set) = self.topic_peers.get(&topic_hash) {
+        let set: Vec<_> = self.peer_topics.keys().collect();
```
Comment (AgeManning) [inline protocols/gossipsub/src/behaviour.rs:664]: This change now gets all peers regardless of which topic they are subscribed to

File: protocols/gossipsub/src/behaviour.rs [hunk 2]
```diff
@@ -675,7 +676,7 @@ where
             } else {
                 // Explicit peers
                 for peer in &self.explicit_peers {
-                    if set.contains(peer) {
+                    if set.contains(&peer) {
```
Comment (AgeManning) [inline protocols/gossipsub/src/behaviour.rs:679]: This is the part that if we are not flood_publishing, then now we send messages to all peers regardless if they are subscribed or not. 

I think this is not what we want to do and is not what the spec is implying.

File: protocols/gossipsub/src/behaviour.rs [hunk 3]
```diff
@@ -1855,7 +1856,9 @@ where
                 "Received message on a topic we are not subscribed to: {:?}",
                 message.topic
             );
-            return;
+            if !self.config.flood_publish() {
```
Comment (AgeManning) [inline protocols/gossipsub/src/behaviour.rs:1859]: If we get a message on a topic that we are not subscribed to, i think we want to just drop it. Its an error and it shouldn't occur if the code is written correctly. We shouldn't then forward this message on, i dont think.

**Review by AgeManning — CHANGES_REQUESTED** [review]

File: protocols/gossipsub/src/behaviour.rs (hunk 2 above)
Comment (bochaco) [reply protocols/gossipsub/src/behaviour.rs:679]: Wouldn't this send the message (if we are not flood_publishing) only to **explicit peers** found in the `set` ?...regardless if they are subscribed or not, i.e. regardless how we calculate the `set` we'd send it to them only if they are set by the user as explicit peers....?

Comment (bochaco) [reply protocols/gossipsub/src/behaviour.rs:679]: ...unless explicit peers are yet required to be subscribed to the topic and only their score is ignored..., is that what explicit peers are meant to be? if so, then yes, I see this is wrong.
I thought they were explicit and not even required to be subscribed to the topic.

File: protocols/gossipsub/src/behaviour.rs (hunk 1 above)
Comment (bochaco) [reply protocols/gossipsub/src/behaviour.rs:664]: Yes, true, I think this is wrong here even for what I was attempting to do.

Basically what I see is that `self.topic_peers.get(&topic_hash)` is `None` when you try to publish a msg from a peer which is not aware of a peer subscribed to that topic. Thus I thought the `flood-publish` is what would enable such a msg in such case to still be sent to connected peers so it eventually finds its way through.

File: protocols/gossipsub/src/behaviour.rs (hunk 2 above)
Comment (AgeManning) [reply protocols/gossipsub/src/behaviour.rs:679]: Ah yep, I didn't look too closely at this. You're right. Its the explicit peers. We still want to segregate them based on topic tho (I gave some reasons in the issue that links to this PR) :)

**Comment by AgeManning** [comment]
@bochaco. I think we should continue this discussion in the issue. I don't immediately see any issues with the code as it stands. I think it is doing what we expect to do, and there is likely is a mismatch between what we expect gossipsub to do. 

Happy to re-open this if we agree things need to be changed tho.

**Comment by bochaco** [comment]
> there is likely is a mismatch between what we expect gossipsub to do.

I totally agreed with that, I was misunderstanding the discussed aspects of the protocol, thanks a lot!

**Pull request closed without merging**